src/context/TodoContext.tsx

```diff
@@ -0,0 +1,194 @@
+import React, {
+  createContext,
+  useMemo,
+  useState,
+} from 'react';
+import * as todoService from '../api/todos';
+import { Todo } from '../types/Todo';
+import { ErrorMessage } from '../types/ErrorMessage';
+import { getActiveTodos, getCompletedTodos } from '../utils/getTodos';
+
+interface Props {
+  children: React.ReactNode
+}
+
+interface TodoContextInterface {
+  todos: Todo[],
+  setTodos: React.Dispatch<React.SetStateAction<Todo[]>>,
+  isLoading: boolean,
+  setIsLoading: (isLoading: boolean) => void,
+  errorMessage: ErrorMessage,
+  setErrorMessage: (error: ErrorMessage) => void,
+  addTodoHandler: (newTodo: Omit<Todo, 'id'>) => Promise<void>,
+  completedTodos: Todo[],
+  activeTodos: Todo[],
+  deleteTodoHandler: (id: number) => void;
+  clearCompleted: () => void;
+  todosIdToDelete: number[],
+  setTodosIdToDelete: (todoIdsToDelete: number[]) => void,
+
+  toggleChangeHandler: (todo: Todo) => void,
+  renameTodoHandler: (todo: Todo, newTodoTitle: string) => Promise<void>,
+
+  tempTodo: Todo | null,
+  setTempTodo: React.Dispatch<React.SetStateAction<Todo | null>>,
+}
+
+const initalContext: TodoContextInterface = {
+  todos: [],
+  setTodos: () => {},
+  isLoading: false,
+  setIsLoading: () => {},
+  errorMessage: ErrorMessage.Default,
+  setErrorMessage: () => {},
+  addTodoHandler: async () => {},
+  completedTodos: [],
+  activeTodos: [],
+  deleteTodoHandler: () => {},
+  clearCompleted: () => {},
+  todosIdToDelete: [],
+  setTodosIdToDelete: () => {},
+
+  toggleChangeHandler: async () => {},
+  renameTodoHandler: async () => {},
+
+  tempTodo: null,
+  setTempTodo: () => {},
+};
+
+export const TodoContext = createContext(initalContext);
+
+export const TodoProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState(ErrorMessage.Default);
+  const [isLoading, setIsLoading] = useState(false);
+  const [todosIdToDelete, setTodosIdToDelete] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const completedTodos = getCompletedTodos(todos);
+  const activeTodos = getActiveTodos(todos);
+
+  const addTodoHandler = (newTodo: Omit<Todo, 'id'>) => {
+    setIsLoading(true);
+
+    return todoService.addTodo(newTodo)
+      .then((createdTodo: Todo) => {
+        setTodos((prevTodos) => [...prevTodos, createdTodo]);
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.AddError);
+        throw new Error();
```

Comment:
https://github.com/mate-academy/react_todo-app-with-api/pull/808#discussion_r1337798371

Reply:
Sorry, i forgot to change error-stuff fixes on local side